### ISSUE TITLE
Strange spans for exported classes with decorators

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#10409]
```ts
@dec export class A {}
```

This produces strange spans:

* `ExportNamedDeclaration`: start 5, end 22
* `Class`: start 0, end 22

[Playground](https://playground.oxc.rs/#eNo9j7tqxEAMRX/FqHaRBNK4SiBsm5C0aeSxvAyMJSNp9oHxv+/MGrvSA+ncexcI0MHHQKGh2yzqTUho1nw2y/rP0IJAt4BmrsXu7HiDzjVTCymy770FmekY7lMvaZ9ckW0UnaAbMRmtLcyoRlqJmJJcf8mz8nd2iwOdMgePUvS291mp3F7oB5XYtm1BVPEnovSHwsZ01DMVZ0D29vL6XjJEk4ROwxeVcIoVX0i7myADnekZkBj7RH+SNdCE86E2RY5j3PWCsKukU7Feny6kvVhJvwHX9QHtD3n1)

(ignore that `Program` start span isn't 0 in Playground. That's an artefact of ESTree conversion - it's 0 in Rust AST, as it should be).

`Class` is a child of `ExportNamedDeclaration` but the child has a span which starts before the parent. This seems odd.

I assume we should consider this to be incorrect?

**Comment by overlookmotel** [comment]
@leaysgur FYI, looks like this is root cause of a bunch of TS-ESTree test fails.

**Comment by leaysgur** [comment]
I've found that TS-ESLint seems to be aware of this: https://github.com/typescript-eslint/typescript-eslint/issues/8769 and refs https://github.com/estree/estree/issues/315 .

**Comment by overlookmotel** [comment]
Ah ha. Thanks for digging up that discussion. OK, it seems everyone has a different view on what's "right".

But from https://github.com/estree/estree/issues/315#issuecomment-2033157062, it sounds like Babel won the argument and TS-ESLint intends to align with Babel's behavior in future.

Babel gives:

* `ExportNamedDeclaration`: start 0, end 22
* `ClassDeclaration`: start 0, end 22
* `Decorator`: start 0, end 4

[AST Explorer - Babel](https://ast-explorer.dev/#eNo1jksOwyAMRK9ieZ0TZFWpV+iuZEHArVJRQHyqVoi7104CGw/zZsANHc740h+dTdpiwQkjG6teybE2rC+WDNA3hlTAOJ0zXKF15RkHxk15AIU51GTo9oukcOb7O9jqWE8Hjq4+N5+F3cWBMQQWbh3fn3k5+7sjYctefWiXafi8gozlbHCITEi6hMRZ8RblO/Y/pbhDrw==)

To clarify the various approaches people are taking:

```ts
// Oxc
@dec export class A {}
^^^^ Decorator
^^^^^^^^^^^^^^^^^^^^^^ Class
     ^^^^^^^^^^^^^^^^^ ExportNamedDeclaration

// TS-ESLint
@dec export class A {}
^^^^ Decorator
            ^^^^^^^^^^ Class
     ^^^^^^^^^^^^^^^^^ ExportNamedDeclaration

// Babel
@dec export class A {}
^^^^ Decorator
^^^^^^^^^^^^^^^^^^^^^^ Class
^^^^^^^^^^^^^^^^^^^^^^ ExportNamedDeclaration
```

Personally, I like Babel's approach. I thought it was a hard rule that a child node cannot have a span which exceeds its parent's span. I don't think there are any other places in AST where that rule is broken, and it feels weird to break it.

The only way to maintain the rule is to do what Babel does - extend the span of both `Class` and `ExportNamedDeclaration` back to cover decorators too. Unfortunately this makes the span of `Class` include the `export` keyword, but in my view the alternative is worse.

@Boshen Could you please make a judgement on this?

**Comment by Boshen** [comment]
Both are reasonable, let's align ts-eslint.

**Comment by overlookmotel** [comment]
#10438 makes spans in our AST follow TS-ESLint:

```ts
@dec export class C {}
     ^^^^^^^^^^^^^^^^^ ExportNamedDeclaration
            ^^^^^^^^^^ Class
^^^^                   Decorator
```

Also:

```ts
@dec export default class {}
     ^^^^^^^^^^^^^^^^^^^^^^^ ExportDefaultDeclaration
                    ^^^^^^^^ Class
^^^^                         Decorator
```
